csrc/backend_ops/ncnn/onnx2ncnn/utils.h

```diff
@@ -0,0 +1,394 @@
+      v.push_back(shape_data[j]);
+    }
+  } else {
+    fprintf(stderr, "Unknown data type %d\n", tp.data_type());
+  }
+
+  return v;
+}
+
+static std::vector<float> get_node_attr_from_input_af(const onnx::TensorProto& tp) {
+  int size = 0;
+
+  std::vector<float> v;
+
+  // float
+  if (tp.data_type() == 1) {
+    const float* shape_data = 0;
+    if (tp.has_raw_data()) {
+      shape_data = (const float*)tp.raw_data().data();
+      size = (int)(tp.raw_data().size() / 4);
+    } else {
+      shape_data = tp.float_data().data();
+      size = tp.float_data_size();
+    }
+    for (int j = 0; j < size; j++) {
+      v.push_back(shape_data[j]);
+    }
+  }
+  // double
+  else if (tp.data_type() == 11) {
+    const double* shape_data = 0;
+    if (tp.has_raw_data()) {
+      shape_data = (const double*)tp.raw_data().data();
+      size = (int)(tp.raw_data().size() / 8);
+    } else {
+      shape_data = tp.double_data().data();
+      size = tp.double_data_size();
+    }
+    for (int j = 0; j < size; j++) {
+      v.push_back((float)shape_data[j]);
+    }
+  } else {
+    fprintf(stderr, "Unknown data type %d\n", tp.data_type());
+  }
+
+  return v;
+}
+
+static int get_tensor_proto_data_size(const onnx::TensorProto& tp) {
+  if (tp.has_raw_data()) {
+    if (tp.data_type() == 1 || tp.data_type() == 6) {
+      const std::string& raw_data = tp.raw_data();
+      int size = (int)raw_data.size() / 4;
+      return size;
+    } else if (tp.data_type() == 7 || tp.data_type() == 11) {
+      const std::string& raw_data = tp.raw_data();
+      int size = (int)raw_data.size() / 8;
+      return size;
+    } else if (tp.data_type() == 9) {
+      const std::string& raw_data = tp.raw_data();
+      return 0;
+    }
+  } else if (tp.data_type() == 1) {
+    return tp.float_data_size();
+  } else if (tp.data_type() == 7) {
+    return tp.int64_data_size();
+  } else if (tp.data_type() == 6) {
+    return tp.int32_data_size();
+  } else if (tp.data_type() == 11) {
+    return tp.double_data_size();
+  }
+
+  return 0;
+}
+
+static void fwrite_tensor_proto_data(const onnx::TensorProto& tp, FILE* bp) {
+  int size = get_tensor_proto_data_size(tp);
+
+  if (tp.has_raw_data()) {
+    const std::string& raw_data = tp.raw_data();
+    fwrite(raw_data.data(), sizeof(float), size, bp);
+  } else if (tp.data_type() == 1) {
+    fwrite(tp.float_data().data(), sizeof(float), size, bp);
+  }
+}
+
+static void fwrite_tensor_proto_data_to_float(const onnx::TensorProto& tp, FILE* bp) {
+  int size = get_tensor_proto_data_size(tp);
+  size_t written_size;
+  if (tp.has_raw_data()) {
+    const std::string& raw_data = tp.raw_data();
+    if (tp.data_type() == 6) {
+      int* intdataptr = (int*)raw_data.data();
+      float* floatdataptr = (float*)malloc(sizeof(float) * size);
+      for (int i = 0; i < size; i++) {
+        floatdataptr[i] = (float)intdataptr[i];
+      }
+      written_size = fwrite(floatdataptr, sizeof(float), size, bp);
```

Comment:
Should we release `floatdataptr` after write?

Reply:
done.